src/libcore/num/dec2flt/mod.rs

```diff
@@ -0,0 +1,228 @@
+
+#![doc(hidden)]
+#![unstable(feature = "dec2flt",
+            reason = "internal routines only exposed for testing")]
+
+use prelude::*;
+use num::ParseFloatError as PFE;
+use num::FloatErrorKind;
+use self::parse::{parse_decimal, Decimal, Sign};
+use self::parse::ParseResult::{self, Valid, ShortcutToInf, ShortcutToZero};
+use self::num::digits_to_big;
+use self::rawfp::RawFloat;
+
+mod algorithm;
+mod table;
+mod num;
+// These two have their own tests.
+pub mod rawfp;
+pub mod parse;
+
+/// Entry point for decimal-to-f32 conversion.
+pub fn to_f32(s: &str) -> Result<f32, PFE> {
+    dec2flt(s)
+}
+
+/// Entry point for decimal-to-f64 conversion.
+pub fn to_f64(s: &str) -> Result<f64, PFE> {
+    dec2flt(s)
+}
+
+/// Split decimal string into sign and the rest, without inspecting or validating the rest.
+fn extract_sign(s: &str) -> (Sign, &str) {
+    match s.as_bytes()[0] {
+        b'+' => (Sign::Positive, &s[1..]),
+        b'-' => (Sign::Negative, &s[1..]),
+        // If the string is invalid, we never use the sign, so we don't need to validate here.
+        _ => (Sign::Positive, s),
+    }
+}
+
+/// Convert a decimal string into a floating point number.
+fn dec2flt<T: RawFloat>(s: &str) -> Result<T, PFE> {
+    if s.is_empty() {
+        return Err(PFE { __kind: FloatErrorKind::Empty });
+    }
+    let (sign, s) = extract_sign(s);
+    let flt = match parse_decimal(s) {
+        Valid(decimal) => try!(convert(decimal)),
+        ShortcutToInf => T::infinity(),
+        ShortcutToZero => T::zero(),
+        ParseResult::Invalid => match s {
+            "inf" => T::infinity(),
+            "NaN" => T::nan(),
+            _ => { return Err(PFE { __kind: FloatErrorKind::Invalid }); }
+        }
+    };
+
+    match sign {
+        Sign::Positive => Ok(flt),
+        Sign::Negative => Ok(-flt),
+    }
+}
+
+/// The main workhorse for the decimal-to-float conversion: Orchestrate all the preprocessing
+/// and figure out which algorithm should do the actual conversion.
+fn convert<T: RawFloat>(mut decimal: Decimal) -> Result<T, PFE> {
+    let mut e = decimal.exp;
+    simplify(&mut decimal, &mut e);
+    if let Some(x) = trivial_cases(&decimal, e) {
+        return Ok(x);
+    }
+    // AlgorithmM and AlgorithmR both compute approximately `f * 10^e`.
+    let max_digits = decimal.integral.len() + decimal.fractional.len() + e.abs() as usize;
+    // Remove/shift out the decimal point.
+    let e = e - decimal.fractional.len() as i64;
+    if let Some(x) = algorithm::fast_path(decimal.integral, decimal.fractional, e) {
+        return Ok(x);
+    }
+    // Big32x40 is limited to 1280 bits, which translates to about 385 decimal digits.
+    // If we exceed this, perhaps while calculating `f * 10^e` in Algorithm R or Algorithm M,
+    // we'll crash. So we error out before getting too close, with a generous safety margin.
+    if max_digits > 375 {
+        return Err(PFE { __kind: FloatErrorKind::Invalid });
```

Comment:
The result of `dec2flt` depends on the order of the input with respect to the numbers half-between floating points. These are numbers of the form `(2^53 + 2u + 1) * 2^-s` for some `0 <= u < 2^52` for normals, and `(2u+1)*(2^-1075)` for denormals. These numbers all become integers when multiplied by `10^1075`, and therefore have less than 1075 nonzero digits (to be more precise, after multiplying by `10^s`, you are left with numbers smaller than `2^54*5^s`, so smaller than `10^[log(2,10)*54 + log(5,10)*1075] < 10^768`, so with no more than 768 nonzero digits) - beyond that last digit the only thing that matters is the tie-breaker sticky bit.

For the record, one such number of the maximum possible length is halfway between the largest subnormal and the smallest normal,
```
2.22507385850720113605740979670913197593481954635164564802342610972482222202107694551652952390813508791414915891303962110687008643869459464552765720740782062174337998814106326732925355228688137214901298112245145188984905722230728525513315575501591439747639798341180199932396254828901710708185069063066665599493827577257201576306269066333264756530000924588831643303777979186961204949739037782970490505108060994073026293712895895000358379996720725430436028407889577179615094551674824347103070260914462157228988025818254518032570701886087211312807951223342628836862232150377566662250398253433597456888442390026549819838548794829220689472168983109969836584681402285424333066033985088644580400103493397042756718644338377048603786162277173854562306587467901408672332763671875E-308
```